subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/graph/builder/ModuleResolveState.java

```diff
@@ -181,12 +181,15 @@ private void changeSelection(ComponentState newSelection) {
         assert this.selected != newSelection;
         assert newSelection.getModule() == this;
 
-        // Remove any outgoing edges for the current selection
-        selected.removeOutgoingEdges();
+        ComponentState oldSelected = selected;
 
+        // set new state before calling 'removeOutgoingEdges()' as it may read the state again in case of recursive dependencies
```

Comment:
Nice catch indeed 👍

Reply:
Unfortunately it is not the right solution.

We need the old behavior for a case like this where the platform is upgraded through one of its memebers:
https://builds.gradle.org/viewLog.html?buildId=30064039&tab=buildResultsDiv&buildTypeId=Gradle_Check_Quick_1_dependencyManagement_3

If we to eagerly discard the old state, no selector for a certain module is left during the platform updating process (because the old version of the platform is discarded before the new one comes in).

In the example above the selector for `org:bar` is given by platform 1 at first. But then org:bar itself is triggering the update to platform 2 and causes the removal of its own selector through the removal of platform 1....

I think this is a use case we want to keep supporting.

I am trying to find a solution where we only catch the self-referencing recursion case and keep everything else working.